nni/contrib/distillation/preprocessor.py

```diff
@@ -0,0 +1,183 @@
+    uidd_kwargs
+        Additional keyword arguments pass to `uid_dataset_cls.__init__`.
+        If there are many extra arguments, recommended to directly set `uid_dataset_cls` to None, and then pass in the wrapped `dataset`.
+    labels_split_fn
+        Please refer `Preprocessor.preprocess_labels`.
+    labels_collate_fn
+        Please refer `Preprocessor.create_replay_dataloader`.
+    """
+    def __init__(self, teacher_predict: Callable[[Any], Any], dataset: Dataset, create_dataloader: Callable[[Dataset], DataLoader],
+                 keep_in_memory: bool = False, cache_folder: str | PathLike | None = None, cache_mode: Literal['pickle', 'hdf5'] = 'pickle',
+                 checkpoint_folder: str | PathLike | None = None, uid_dataset_cls: Type[_UidDataset] | None = None,
+                 uidd_args: List | None = None, uidd_kwargs: Dict | None = None, labels_split_fn: Callable[[Any], List] | None = None,
+                 labels_collate_fn: Callable[[List], Any] | None = None):
+        self._storage = create_storage(keep_in_memory=keep_in_memory, cache_folder=cache_folder, cache_mode=cache_mode,
+                                       checkpoint_folder=checkpoint_folder)
+
+        self._teacher_predict = teacher_predict
+        self._dataset = create_uid_dataset(dataset, uid_dataset_cls, uidd_args=uidd_args, uidd_kwargs=uidd_kwargs)
+        self._create_dataloader = create_dataloader
+
+        assert (labels_split_fn is None and labels_collate_fn is None) or (labels_split_fn is not None and labels_collate_fn is not None)
+        self._labels_split_fn = labels_split_fn if labels_split_fn is not None else _default_labels_split_fn
+        self._labels_collate_fn = labels_collate_fn if labels_collate_fn is not None else _default_labels_collate_fn
+
+        self._preprocessed = False if checkpoint_folder is None else True
+
+    def _patched_uid_dataloader(self):
+        dataloader = self._create_dataloader(self._dataset)
+        origin_collate_fn = dataloader.collate_fn
+
+        def collate_fn(data):
+            uids, origin_data = list(zip(*data))
+            return uids, origin_collate_fn(origin_data)  # type: ignore
+
+        setattr(dataloader, 'collate_fn', collate_fn)
+        return dataloader
+
+    def _patched_label_dataloader(self, labels_collate_fn: Callable[[List], Any]):
+        dataloader = self._patched_uid_dataloader()
+
+        uid_collate_fn = dataloader.collate_fn
+
+        def collate_fn(data):
+            uids, origin_batch = uid_collate_fn(data)
+            distil_labels = list(map(self._storage.select, cast(List[str], uids)))
+            distil_labels = labels_collate_fn(distil_labels)
+            return distil_labels, origin_batch
+
+        setattr(dataloader, 'collate_fn', collate_fn)
+        return dataloader
+
+    def preprocess_labels(self, epochs: int = 1, log_process: bool = True, labels_split_fn: Callable[[Any], List] | None = None):
```

Comment:
Did you discuss about renaming this?

Reply:
`generate_distillation_labels`? I will rename it.
Do you think cls name `Preprocessor` should be renamed? Maybe `DistillationLabelGenerator`?